gen/exhibitor_tls_bootstrap.py

```diff
@@ -0,0 +1,121 @@
+import json
+import os
+import pathlib
+import shutil
+import subprocess
+import tempfile
+
+from typing import Any, Dict, List
+from urllib.parse import urlparse
+
+PACKAGE_NAME = 'dcoscertstrap'
+BINARY_PATH = '/genconf/bin'
+CA_PATH = '/genconf/ca'
+
+
+def _check(config: Dict[str, Any]) -> List[str]:
+    """
+    Current constraints are:
+        config.yaml:
+            exhibitor_tls_enabled == True,
+            bootstrap_url must not be a local path
+                This is necessary to prevent this orchestration
+                from running when gen is not executed on a proper
+                bootstrap node.
+            master_discovery must be static
+         DC/OS variant must be enterprise
+    """
+    checks = [
+        (lambda: config.get('exhibitor_tls_enabled', False) == 'true',
+         'Exhibitor security is disabled'),
+        (lambda: config['master_discovery'] == 'static',
+         'Only static master discovery is supported'),
+        (lambda: config['dcos_variant'] == 'enterprise',
+         'Exhibitor security is an enterprise feature'),
+        (lambda: urlparse(config['bootstrap_url']).scheme != 'file',
+         'Blackbox exhibitor security is only supported when using a remote'
+         ' bootstrap node'),
+    ]
+
+    reasons = []
+    for check in checks:
+        if not check[0]():
+            reasons.append(check[1])
+
+    return reasons
+
+
+def find_package(packages_json: str) -> str:
+    packages = json.loads(packages_json)
+    for package in packages:
+        if package.startswith(PACKAGE_NAME):
+            return package
+    raise Exception('dcoscertstrap package is not present')
+
+
+def _extract_package(package_path: str):
+    """ Extracts the dcoscertstrap package from the local pkgpanda
+    repository """
+    os.makedirs(BINARY_PATH, exist_ok=True)
+    with tempfile.TemporaryDirectory() as td:
+        subprocess.run(['tar', '-xJf', package_path, '-C', td])
+
+        shutil.move(
+            pathlib.Path(td) / 'bin' / PACKAGE_NAME,
+            pathlib.Path(BINARY_PATH) / PACKAGE_NAME)
+
+
+def _init_ca(alt_names: List[str]):
+    """ Initializes the CA (generates a private key and self signed
+    certificate CA extensions) if the resulting files do not already exist """
+    if not all(map(os.path.exists, [
+            pathlib.Path(CA_PATH) / 'root-cert.pem',
+            pathlib.Path(CA_PATH) / 'root-key.pem'])):
```

Comment:
Hmm, I would go with either `os.path` or `pathlib` functions but not both at the same time.